packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp

```diff
@@ -187,6 +177,91 @@ void jacobiIterate(RCP<Teuchos::ParameterList> smootherParams,
   return;
 } // jacobiIterate
 
+/*! \brief Do Gauss-Seidel smoothing
+ *
+ *  Perform Gauss-Seidel smoothing in the region layout using the true diagonal value
+ *  recovered from the splitted matrix.
+ */
+template<class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+void GSIterate(RCP<Teuchos::ParameterList> smootherParams,
+                   std::vector<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > >& regX, // left-hand side (or solution)
+                   const std::vector<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regB, // right-hand side (or residual)
+                   const std::vector<RCP<Xpetra::Matrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regionGrpMats, // matrices in true region layout
+                   const std::vector<RCP<Xpetra::Vector<Scalar, LocalOrdinal, GlobalOrdinal, Node> > > regionInterfaceScaling, // recreate on coarse grid by import Add on region vector of ones
+                   const int maxRegPerProc, ///< max number of regions per proc [in]
+                   const RCP<Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> > mapComp, ///< composite map
+                   const std::vector<RCP<Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> > > rowMapPerGrp, ///< row maps in region layout [in] requires the mapping of GIDs on fine mesh to "filter GIDs"
+                   const std::vector<RCP<Xpetra::Map<LocalOrdinal, GlobalOrdinal, Node> > > revisedRowMapPerGrp, ///< revised row maps in region layout [in] (actually extracted from regionGrpMats)
+                   const std::vector<RCP<Xpetra::Import<LocalOrdinal, GlobalOrdinal, Node> > > rowImportPerGrp ///< row importer in region layout [in]
+    )
+{
+#include "Xpetra_UseShortNames.hpp"
+  const Scalar SC_ZERO = Teuchos::ScalarTraits<Scalar>::zero();
+  const Scalar SC_ONE = Teuchos::ScalarTraits<Scalar>::one();
+  const int maxIter    = smootherParams->get<int>   ("smoother: sweeps");
+  const double damping = smootherParams->get<double>("smoother: damping");
+
+
+  std::vector<RCP<Vector> > regRes(maxRegPerProc);
+  createRegionalVector(regRes, maxRegPerProc, revisedRowMapPerGrp);
+
+  // extract diagonal from region matrices, recover true diagonal values, invert diagonal
```

Comment:
I think the extraction of the diagonal should go in a setup phase and simply be reused here instead of recomputing it on the fly.
That being said I don't mind if that change happens in another PR as it is not essential for a first cut.